lib/c.c

```diff
@@ -346,23 +346,88 @@ void printf(char *str, ...)
                 /* append literal '%' character */
                 buffer[bi] = '%';
                 bi++;
-                pi++;
+                si++;
+                continue;
             }
-            pi--;
+            pi++;
             si++;
         }
     }
     buffer[bi] = 0;
     __syscall(__syscall_write, 1, buffer, bi);
 }
 
+void sprintf(char *buffer, char *str, ...)
+{
+    int *var_args = &str + 4;
+    int si = 0, bi = 0, pi = 0;
+
+    while (str[si]) {
+        if (str[si] != '%') {
+            buffer[bi] = str[si];
+            bi++;
+            si++;
+        } else {
+            int w = 0, zp = 0, pp = 0;
+
+            si++;
+            if (str[si] == '#') {
+                pp = 1;
+                si++;
+            }
+            if (str[si] == '0') {
+                zp = 1;
+                si++;
+            }
+            if (str[si] >= '1' && str[si] <= '9') {
+                w = str[si] - '0';
+                si++;
+                if (str[si] >= '0' && str[si] <= '9') {
+                    w = w * 10;
+                    w += str[si] - '0';
+                    si++;
+                }
+            }
+            if (str[si] == 's') {
```

Comment:
Use switch-case for readable code.

Reply:
May I change it with `printf()` together in another pull request?

Reply:
For newly-added code, it's better to make improvements sooner rather than requiring extra effort later on.